### PR TITLE
SDK-437 [Node.js SDK] Add microseconds support

### PR DESCRIPTION
**Pull request by igor-alekseev-dev**
Pull Request Checklist
======================

Please read this for our PR Culture:
------------------------------------
https://incountry.atlassian.net/wiki/spaces/ED/pages/46858869/Pull+requests+culture+Draft

Work relates to:
----------------
Link(s) to Jira Task/User Story/Bug/Epic:
- [SDK-437](https://incountry.atlassian.net/browse/SDK-437)

Code Quality (Before submitting PR)
------------
- [x] Ensure code compiles and passes Sonar Lint.
- [x] Do your best to ensure integration tests pass. If they don't then...
- [x] Remove, ignore, or mark pending any brittle tests.
- [x] Remove commented out code.

Target Release
--------------
- [ ] Please put the target release vehicle this needs to be bundled with.
- [ ] Did you increment the version in package.json? What is the new version?

### CODE REVIEW

**Review by artemgolozubov — COMMENTED** [review]
Please fix test names. Also _underscores _in _storage _looks _ugly, _but _I'm not sure we need to fix this at the moment. And please take a look at the rest of comments.

File: README.md [hunk 1]
```diff
@@ -298,10 +298,10 @@ type StorageRecordData = {
   serviceKey5?: string | null;
   body?: string | null;
   precommitBody?: string | null;
-  rangeKey1?: t.Int | null;
+  rangeKey1?: number | null; // Accepts only Integer numbers
   ...,
-  rangeKey10?: t.Int | null;
-  expiresAt?: string | Date | null; // Accepts only ISO-8601 formatted strings, like '2021-03-11T17:23:05.941Z'
+  rangeKey10?: number | null;
+  expiresAt?: string | number | Date | MicroDate | null; // Accepts ISO-8601 formatted strings with microseconds, number timestamp with decimal microseconds, Date and MicroDate
```
Comment (RommeroInCountry) [inline README.md:304]: number timestamp -> numeric timestamps
Date and MicroDate -> Date and MicroDate objects

File: README.md [hunk 2]
```diff
@@ -426,7 +426,8 @@ const writeResult = await storage.write(countryCode, recordData);
 
 All dates are stored in UTC timezone and converted to UTC timezone in case of `expiresAt`.
 
-SDK accepts ISO-8601 formatted strings, e.g '2021-03-11T17:23:05.941Z'.
+SDK accepts ISO-8601 formatted strings, e.g '2021-03-11T17:23:05.941Z', number timestamp with decimal microseconds, Date and MicroDate.
```
Comment (RommeroInCountry) [inline README.md:429]: number timestamp -> numeric timestamps
Date and MicroDate -> Date and MicroDate objects

File: src/validation/date.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import * as t from 'io-ts';
+import { MicroDate } from '../micro-date';
+import { codecFromIs } from './utils';
+
+const isDate = (o: unknown): o is Date => o instanceof Date;
+const DateIO = codecFromIs('Date', isDate);
+
+const DATE_ERROR_MALFORMED_ISO_STRING = 'Malformed DateTime ISO string';
+const DATE_ERROR_CANNOT_CREATE_MICRODATE = 'Cannot create MicroDate: Expected string, number, Date or MicroDate';
```
Comment (RommeroInCountry) [inline src/validation/date.ts:9]: We can go further and add `Cannot create MicroDate: Expected string, number, Date or MicroDate. Got: ${value}`

File: tests/unit/storage/add-attachment-tests.ts [hunk 4]
```diff
@@ -68,7 +68,7 @@ describe('Storage', () => {
             const wrongCountries = [undefined, null, 1, {}, []];
             // @ts-ignore
             await Promise.all(wrongCountries.map((country) => expect(encStorage.addAttachment(country))
-              .to.be.rejectedWith(InputValidationError, errorMessageRegExp('addAttachment() Validation Error:', COUNTRY_CODE_ERROR_MESSAGE))));
+              .to.be.rejectedWith(InputValidationError, errorMessageRegExp('addAttachment() call: Validation Error:', COUNTRY_CODE_ERROR_MESSAGE))));
```
Comment (RommeroInCountry) [inline tests/unit/storage/add-attachment-tests.ts:71]: why and where this "call:" was introduced to the error stack trace?

**Review by RommeroInCountry — CHANGES_REQUESTED** [review]
We need an integration test suite covering the following case:
1. Create a record
2. Extract created_at from the record (be making a read request)
3. Make a find request using created_at value
4. Make sure find properly returns the expected record
5. Do the very same thing with updated_at and expires_at with microseconds

File: tests/unit/storage/add-attachment-tests.ts (hunk 4 above)
Comment (igor-alekseev-dev) [reply tests/unit/storage/add-attachment-tests.ts:71]: Former error message structure `Storage.addAttachment() call: addAttachment() Validation Error: aaa`
- `addAttachment() call:` added by `@nomalizeErrors` decorator 
- `addAttachment() Validation Error:` added by `@validate` decorator.



Current `addAttachment() call: Validation Error: aaa`. 

So error message became shorter and changed a little.

File: src/validation/date.ts (hunk 3 above)
Comment (igor-alekseev-dev) [reply src/validation/date.ts:9]: Received value is logged before this message

**Review by RommeroInCountry — DISMISSED** [review]

**Review by RommeroInCountry — APPROVED** [review]

**Review by artemgolozubov — APPROVED** [review]
LGTM

**Pull request merged**